### PR TITLE
Move nav bar data model to be pure data

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Initial step prio to moving nav bar computation out of process.

### CODE REVIEW

File: src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicNavigationBarItemService_CodeGeneration.vb [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+Imports Microsoft.CodeAnalysis.Simplification
+Imports Microsoft.CodeAnalysis.Text
+Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
+Imports Microsoft.VisualStudio.Text.Editor
+
+Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.NavigationBar
+    Partial Friend Class VisualBasicNavigationBarItemService
+        Inherits AbstractNavigationBarItemService
+
+        Private Sub GenerateCodeForItem(document As Document, generateCodeItem As AbstractGenerateCodeItem, textView As ITextView, cancellationToken As CancellationToken)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicNavigationBarItemService_CodeGeneration.vb:25]: all these methods moved here from teh individual RoslynNavigationBArItem subclasses.

File: src/EditorFeatures/Core/Extensibility/NavigationBar/NavigationBarItems/RoslynNavigationBarItem.AbstractGenerateCodeItem.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Extensibility.NavigationBar
+{
+    internal abstract partial class RoslynNavigationBarItem
+    {
+        public abstract class AbstractGenerateCodeItem : RoslynNavigationBarItem
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Extensibility/NavigationBar/NavigationBarItems/RoslynNavigationBarItem.AbstractGenerateCodeItem.cs:12]: these data types moved up from VB.  tehy moved to be pure data.  All logic in tehm moved into the INavigationBarItemService.

**Review by genlu — APPROVED** [review]
:shipit:

File: src/EditorFeatures/Core/Extensibility/NavigationBar/NavigationBarItems/RoslynNavigationBarItem.GenerateDefaultConstructorItem.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.CodeAnalysis.Editor.Extensibility.NavigationBar
+{
+    internal abstract partial class RoslynNavigationBarItem
+    {
+        public class GenerateDefaultConstructor : AbstractGenerateCodeItem
+        {
+            public GenerateDefaultConstructor(string text, SymbolKey destinationTypeSymbolKey)
+                : base(RoslynNavigationBarItemKind.GenerateDefaultConstructor, text, Glyph.MethodPublic, destinationTypeSymbolKey)
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Extensibility/NavigationBar/NavigationBarItems/RoslynNavigationBarItem.GenerateDefaultConstructorItem.cs:12]: Are the kinds for a follow-up where we're actually serializing this?

File: src/EditorFeatures/Core/Extensibility/NavigationBar/NavigationBarItems/RoslynNavigationBarItem.GenerateEventHandlerItem.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.CodeAnalysis.Editor.Extensibility.NavigationBar
+{
+    internal abstract partial class RoslynNavigationBarItem
+    {
+        public class GenerateEventHandler : AbstractGenerateCodeItem
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Extensibility/NavigationBar/NavigationBarItems/RoslynNavigationBarItem.GenerateEventHandlerItem.cs:9]: Why did these have to move out of the VB code? My "why" question here isn't that this is _bad_ just not sure where you're going with this.

File: src/EditorFeatures/Core/Extensibility/NavigationBar/NavigationBarItems/RoslynNavigationBarItemKind.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+namespace Microsoft.CodeAnalysis.Editor.Extensibility.NavigationBar
+{
+    internal enum RoslynNavigationBarItemKind
+    {
+        Symbol,
+        GenerateDefaultConstructor,
+        GenerateEventHandler,
+        GenerateFinalizer,
+        GenerateMethod,
+    }
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Extensibility/NavigationBar/NavigationBarItems/RoslynNavigationBarItemKind.cs:14]: Just asking to understand future intent: what is this being used for? I assume it's for something other than just the type switching in the generate method?

File: src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicNavigationBarItemService.vb [hunk 6]
```diff
@@ -2,16 +2,16 @@
 ' The .NET Foundation licenses this file to you under the MIT license.
 ' See the LICENSE file in the project root for more information.
 
+Imports System.Collections.Immutable
 Imports System.Composition
 Imports System.Threading
-Imports System.Threading.Tasks
 Imports Microsoft.CodeAnalysis
 Imports Microsoft.CodeAnalysis.Editor.Extensibility.NavigationBar
+Imports Microsoft.CodeAnalysis.Editor.Extensibility.NavigationBar.RoslynNavigationBarItem
```
Comment (jasonmalinowski) [inline src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicNavigationBarItemService.vb:10]: Is this a namespace or class being imported here?

File: src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicNavigationBarItemService.vb [hunk 7]
```diff
@@ -423,31 +426,31 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.NavigationBar
...
-                        symbolIdIndexProvider.GetIndexForSymbolId(method.GetSymbolKey()),
+                        method.GetSymbolKey(cancellationToken),
+                        symbolIdIndexProvider.GetIndexForSymbolId(method.GetSymbolKey(cancellationToken)),
                         bolded:=spans.Count > 0,
                         grayed:=spans.Count = 0))
                 ElseIf method IsNot Nothing AndAlso IsUnimplementedPartial(method) Then
                     If workspaceSupportsDocumentChanges Then
-                        items.Add(New GenerateMethodItem(
+                        items.Add(New GenerateMethod(
                         member.ToDisplayString(displayFormat),
```
Comment (jasonmalinowski) [inline src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicNavigationBarItemService.vb:440]: Looks like indenting is screwed up here (you were fixing others so I'll assume this is fair game!)

**Review by jasonmalinowski — APPROVED** [review]
Some questions but they're most of the "I presume you have a plan for this bit otherwise I'm a touch confused."

File: src/EditorFeatures/Core/Extensibility/NavigationBar/NavigationBarItems/RoslynNavigationBarItem.GenerateDefaultConstructorItem.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Extensibility/NavigationBar/NavigationBarItems/RoslynNavigationBarItem.GenerateDefaultConstructorItem.cs:12]: yes.

File: src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicNavigationBarItemService.vb (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicNavigationBarItemService.vb:10]: a class.

File: src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicNavigationBarItemService.vb (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicNavigationBarItemService.vb:440]: yes :)

**Pull request merged**